habitat-lab/habitat/core/embodied_task.py

```diff
@@ -295,45 +295,43 @@ def reset(self, episode: Episode):
 
     def _step_single_action(
         self,
-        observations: Any,
         action_name: Any,
         action: Dict[str, Any],
         episode: Episode,
-        is_last_action=True,
     ):
         if isinstance(action_name, (int, np.integer)):
             action_name = self.get_action_name(action_name)
         assert (
             action_name in self.actions
         ), f"Can't find '{action_name}' action in {self.actions.keys()}."
         task_action = self.actions[action_name]
-        observations.update(
-            task_action.step(
-                **action["action_args"],
-                task=self,
-                is_last_action=is_last_action,
-            )
+        return task_action.step(
+            **action["action_args"],
+            task=self,
         )
 
     def step(self, action: Dict[str, Any], episode: Episode):
         action_name = action["action"]
         if "action_args" not in action or action["action_args"] is None:
             action["action_args"] = {}
-        observations: Any = {}
+        observations: Any = None
         if isinstance(action_name, tuple):  # there are multiple actions
-            for i, a_name in enumerate(action_name):
-                self._step_single_action(
-                    observations,
+            for a_name in action_name:
+                observations = self._step_single_action(
                     a_name,
                     action,
                     episode,
-                    i == len(action_name) - 1,
                 )
         else:
-            self._step_single_action(
-                observations, action_name, action, episode
+            observations = self._step_single_action(
+                action_name, action, episode
             )
```

Comment:
After this code, if observations is None, then we will need to get the observations from the simulator and lab sensors. If it is not None, this means the environment was stepped by the action, so we DO NOT call sim.step()